walnuts_cpp/include/walnuts/nuts.hpp

```diff
@@ -154,36 +161,30 @@ Span<S> build_leaf(const F &logp_grad_fun, const Span<S> &span,
 }
 
 template <Direction D, typename S, class F, class Generator>
-Span<S> build_span(Random<S, Generator> &rng, const F &logp_grad_fun,
-                   const Vec<S> &inv_mass, S step, Integer depth,
-                   const Span<S> &last_span, bool &uturn_flag) {
-  uturn_flag = false;
+std::optional<Span<S>> build_span(Random<S, Generator> &rng,
+                                  const F &logp_grad_fun,
+                                  const Vec<S> &inv_mass, S step, Integer depth,
+                                  const Span<S> &last_span) {
   if (depth == 0) {
     return build_leaf<D>(logp_grad_fun, last_span, inv_mass, step);
   }
-  Span<S> span1 = build_span<D>(rng, logp_grad_fun, inv_mass, step, depth - 1,
-                                last_span, uturn_flag);
-  if (uturn_flag) {
-    return last_span; // won't be used
+  auto span1 =
+      build_span<D>(rng, logp_grad_fun, inv_mass, step, depth - 1, last_span);
+  if (!span1) {
+    return std::nullopt;
   }
-  Span<S> span2 = build_span<D>(rng, logp_grad_fun, inv_mass, step, depth - 1,
-                                span1, uturn_flag);
-  if (uturn_flag) {
-    return last_span; // won't be used
+
+  auto span2 =
+      build_span<D>(rng, logp_grad_fun, inv_mass, step, depth - 1, *span1);
+  if (!span2) {
+    return std::nullopt;
   }
-  if constexpr (D == Direction::Forward) {
-    if (uturn(span1, span2, inv_mass)) {
-      uturn_flag = true;
-      return last_span; // won't be used
-    }
-  } else { // Direction::Backward
-    if (uturn(span2, span1, inv_mass)) {
-      uturn_flag = true;
-      return last_span; // won't be used
-    }
+
+  if (uturn<D>(*span1, *span2, inv_mass)) {
+    return std::nullopt;
   }
-  return combine<Update::Barker, D>(rng, std::move(span1), std::move(span2),
-                                    inv_mass, uturn_flag);
+
+  return combine<Update::Barker, D>(rng, *std::move(span1), *std::move(span2));
```

Comment:
Why is this `*std::move(span1)` rather than `std::move(*span1)`?   According to GPT, the former only returns an rvalue reference as of C++17 (which should be OK if we insist on >= C++17) and the latter is more idiomatic.  

StackOverflow concurs:  https://stackoverflow.com/questions/72088667/should-i-move-the-value-out-of-an-optional-or-move-the-whole-optional

Reply:
I also think the code will be a bit clearer if we don't use variables named `span` for both optional and non-optional spans.  How about `ospan` or `opt_span` for the optional ones?  I don't usually want to build types into variable names, but I feel the distinction's a bit confusing here and don't know how to call it out better.

Reply:
std::optional only exists in C++17, so I'm not sure what GPT is thinking it would do before that...

I'm happy to go either way on this. The SO post mentions that it might be hard to tell what is happening during an assignment, which I would agree, but this is during a function call that will fail to typecheck if given anything other than an rvalue ref, so it should be unambiguous that it is doing the correct thing here. 

I went with `maybe_...` for the names, as I think that's a bit more semantic. Let me know what you think